crates/bevy_input_focus/src/directional_navigation.rs

```diff
@@ -0,0 +1,379 @@
+    ///
+    /// If no neighbor exists in a given direction, the value will be [`None`].
+    /// In most cases, using [`NavNeighbors::set`] and [`NavNeighbors::get`]
+    /// will be more ergonomic than directly accessing this array.
+    pub neighbors: [Option<Entity>; 8],
+}
+
+impl NavNeighbors {
+    /// An empty set of neighbors.
+    pub const EMPTY: NavNeighbors = NavNeighbors {
+        neighbors: [None; 8],
+    };
+
+    /// Get the neighbor for a given [`CompassOctant`].
+    pub const fn get(&self, octant: CompassOctant) -> Option<Entity> {
+        self.neighbors[octant.to_index()]
+    }
+
+    /// Set the neighbor for a given [`CompassOctant`].
+    pub const fn set(&mut self, octant: CompassOctant, entity: Entity) {
+        self.neighbors[octant.to_index()] = Some(entity);
+    }
+}
+
+/// A resource that stores the traversable graph of focusable entities.
+///
+/// Each entity can have up to 8 neighbors, one for each [`CompassOctant`].
+///
+/// To ensure that your graph is intuitive to navigate and generally works correctly, it should be:
+///
+/// - **Connected**: Every focusable entity should be reachable from every other focusable entity.
+/// - **Symmetric**: If entity A is a neighbor of entity B, then entity B should be a neighbor of entity A, ideally in the reverse direction.
+/// - **Physical**: The direction of navigation should match the layout of the entities when possible,
+///   although looping around the edges of the screen is also acceptable.
+/// - **Not self-connected**: An entity should not be a neighbor of itself; use [`None`] instead.
+///
+/// For now, this graph must be built manually, and the developer is responsible for ensuring that it meets the above criteria.
+#[derive(Resource, Debug, Default, Clone, PartialEq)]
+pub struct DirectionalNavigationMap {
+    /// A directed graph of focusable entities.
+    ///
+    /// Pass in the current focus as a key, and get back a collection of up to 8 neighbors,
+    /// each keyed by a [`CompassOctant`].
+    pub neighbors: EntityHashMap<NavNeighbors>,
+}
+
+impl DirectionalNavigationMap {
+    /// Adds a new entity to the navigation map, overwriting any existing neighbors for that entity.
+    ///
+    /// Removes an entity from the navigation map, including all connections to and from it.
+    ///
+    /// Note that this is an O(n) operation, where n is the number of entities in the map,
+    /// as we must iterate over each entity to check for connections to the removed entity.
+    ///
+    /// If you are removing multiple entities, consider using [`remove_multiple`](Self::remove_multiple) instead.
+    pub fn remove(&mut self, entity: Entity) {
+        self.neighbors.remove(&entity);
+
+        for node in self.neighbors.values_mut() {
+            for neighbor in node.neighbors.iter_mut() {
+                if *neighbor == Some(entity) {
+                    *neighbor = None;
+                }
+            }
+        }
```

Comment:
If we forced the graph to be symmetric we could make this significantly more efficient (and constant-time bounded).

Reply:
Yeah, sadly I don't think that's viable. As discussed on Discord, the "three small elements beside one big one" layout is not symmetric.